### PR TITLE
Modernize ranges::swap_ranges and ranges::distance

### PR DESCRIPTION
**Pull request by miscco**
This *swaps* the implementation of the the last two algorithms to bridge the *distance* to the other ones.

### CODE REVIEW

**Comment by CaseyCarter** [comment]
> This _swaps_ the implementation of the the last two algorithms to bridge the _distance_ to the other ones.

I think I'm having a negative influence on you ;)

File: stl/inc/xutility [hunk 1]
```diff
@@ -3225,25 +3225,33 @@ namespace ranges {
...
                 _Adl_verify_range(_First, _Last);
-                auto _UFirst                  = _Get_unwrapped(static_cast<_It&&>(_First));
-                const auto _ULast             = _Get_unwrapped(static_cast<_Se&&>(_Last));
-                iter_difference_t<_It> _Count = 0;
-                for (; _UFirst != _ULast; ++_UFirst) {
-                    ++_Count;
-                }
-
-                return _Count;
+                return static_cast<iter_difference_t<_It>>(
```
Comment (CaseyCarter) [inline stl/inc/xutility:3228]: There are a great many places in the STL that assume that corresponding wrapped-and-unwrapped iterator types have the same associated types. This `static_cast` (ditto 3238) should therefore be unnecessary. If you think we should be paranoid - I'm not bothered - I'd prefer to `static_assert(is_same_v<iter_difference_t<_It>, iter_difference_t<_Unwrapped_t<_It>>>, "YOU NO DO THIS THING!");`.

Comment (BillyONeal) [reply stl/inc/xutility:3228]: I would not like to see a static_assert because vector<T, small_allocator<T>> fails your test (unwraps to T*), but I also don't mind generating tons of warnings for this case so I would still like to see the static assert be gone.

Comment (miscco) [reply stl/inc/xutility:3228]: Did I parse that as "No change needed"?

**Comment by miscco** [comment]
BTW can somebody tell my why this changes, because it totally works on my machine

**Comment by statementreply** [comment]
> why does it not swap the ranges?
There is a bug in ranges::iter_swap (#1067).

**Comment by miscco** [comment]
> > why does it not swap the ranges?
> 
> There is a bug in `ranges::iter_swap` (#1067).

Hooray for increased test coverage

File: stl/inc/xutility (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/xutility:3228]: My comment: "we shouldn't cast here; you can `static_assert` if you feel it's necessary"
Billy's comment: "we shouldn't cast _or_ `static_assert` here" (The final occurrence of "static assert" is intended to be "static cast".)

TLDR: let's remove the cast.

**Comment by CaseyCarter** [comment]
@miscco, could you verify that #1072 fixes this?

**Comment by miscco** [comment]
At least on my machine it works now, but it did before so ...

**Comment by statementreply** [comment]
```
stl-lit.py: C:\Users\He\source\repos\STL\tests\utils\stl\test\format.py:203: note: Build step failed unexpectedly.
Command: "C:\Program Files\LLVM\bin\clang-cl.EXE" "/FeC:\Users\He\source\repos\STL\out\build\x64\tests\std\tests\P0896R4_ranges_alg_swap_ranges\15\P0896R4_ranges_alg_swap_ranges.exe" "/FoC:\Users\He\source\repos\STL\out\build\x64\tests\std\tests\P0896R4_ranges_alg_swap_ranges\15\P0896R4_ranges_alg_swap_ranges.obj" "/IC:\Users\He\source\repos\STL\out\build\x64\out\inc" "/IC:/Users/He/source/repos/STL/llvm-project/libcxx/test/support" "/IC:/Users/He/source/repos/STL/tests/std/include" "-m64" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/sdl" "/WX" "/Zc:strictStrings" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/w14640" "/Zc:threadSafeInit-" "/EHsc" "/std:c++latest" "-fno-ms-compatibility" "-fno-delayed-template-parsing" "/MD" "C:\Users\He\source\repos\STL\tests\std\tests\P0896R4_ranges_alg_swap_ranges\test.cpp" "/link" "/LIBPATH:C:\Users\He\source\repos\STL\out\build\x64\out\lib\amd64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29109\lib\x64" "/MANIFEST:EMBED"
Exit Code: 1
Standard Error:
--
C:\Users\He\source\repos\STL\tests\std\tests\P0896R4_ranges_alg_swap_ranges\test.cpp(67,19): error: static_assert expression is not an integral constant expression
    STATIC_ASSERT((test_in_in<instantiator, int, int>(), true));
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/Users/He/source/repos/STL/tests/std/include\range_algorithm_support.hpp(15,42): note: expanded from macro 'STATIC_ASSERT'
#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
                                         ^~~~~~~~~~~
C:\Users\He\source\repos\STL\out\build\x64\out\inc\xutility(682,111): note: constexpr evaluation hit maximum step limit; possible infinite loop?
            _NODISCARD constexpr decltype(auto) operator()(_Ty&& _Val) const noexcept(_Choice<_Ty>._No_throw) {
                                                                                                              ^
C:\Users\He\source\repos\STL\out\build\x64\out\inc\xutility(1033,57): note: in call to '&iter_move->operator()(_First2)'
            iter_value_t<remove_reference_t<_Xty>> _Tmp(_RANGES iter_move(_XVal));
                                                        ^
C:\Users\He\source\repos\STL\out\build\x64\out\inc\yvals_core.h(1218,20): note: expanded from macro '_RANGES'
#define _RANGES    ::std::ranges::
                   ^
C:\Users\He\source\repos\STL\out\build\x64\out\inc\xutility(1071,25): note: in call to '_Iter_exchange_move(_First2, _First1)'
                        _Iter_exchange_move(static_cast<_Ty2&&>(_Val2), static_cast<_Ty1&&>(_Val1));
                        ^
C:\Users\He\source\repos\STL\out\build\x64\out\inc\algorithm(3273,17): note: in call to '&iter_swap->operator()(_First1, _First2)'
                _RANGES iter_swap(_First1, _First2);
                ^
C:\Users\He\source\repos\STL\out\build\x64\out\inc\yvals_core.h(1218,20): note: expanded from macro '_RANGES'
#define _RANGES    ::std::ranges::
                   ^
C:\Users\He\source\repos\STL\out\build\x64\out\inc\algorithm(3254,29): note: in call to '_Swap_ranges_unchecked({&input1[0]}, {&input1[3]}, {&input2[0]}, {&input2[3]})'
            auto _UResult = _Swap_ranges_unchecked(
                            ^
C:\Users\He\source\repos\STL\tests\std\tests\P0896R4_ranges_alg_swap_ranges\test.cpp(54,27): note: in call to '&swap_ranges->operator()(wrapped_input1, wrapped_input2)'
            auto result = swap_ranges(wrapped_input1, wrapped_input2);
                          ^
C:/Users/He/source/repos/STL/tests/std/include\range_algorithm_support.hpp(877,9): note: in call to 'call()'
        Continuation::template call<Args...,
        ^
C:/Users/He/source/repos/STL/tests/std/include\range_algorithm_support.hpp(904,9): note: in call to 'call()'
        Continuation::template call<Args...,
        ^
C:/Users/He/source/repos/STL/tests/std/include\range_algorithm_support.hpp(1038,5): note: in call to 'call()'
    with_input_ranges<with_input_ranges<Instantiator, Element2>, Element1>::call();
    ^
C:\Users\He\source\repos\STL\tests\std\tests\P0896R4_ranges_alg_swap_ranges\test.cpp(67,20): note: in call to 'test_in_in()'
    STATIC_ASSERT((test_in_in<instantiator, int, int>(), true));
                   ^
1 error generated.
```

clang ran out of constexpr evaluation limit on my machine.

**Comment by CaseyCarter** [comment]
This is no longer blocked on #1062: we should be able to merge master and get this cleaned up and across the finish line.

File: tests/std/tests/P0896R4_ranges_alg_swap_ranges/test.cpp [hunk 2]
```diff
@@ -2,65 +2,105 @@
...
+
+#ifdef TEST_EVERYTHING
+int main() {
+    // no constexpr tests here; the below exceeds constexpr step limits
+    test_in_in<instantiator, int, int>();
+}
+#else // ^^^ test all permutations of range properties / test only interesting permutations vvv
+template <class Category, test::ProxyRef IsProxyRef>
+using test_range = test::range<Category, int, test::Sized::no, test::CanDifference::no, test::Common::no,
+    test::CanCompare{derived_from<Category, std::forward_iterator_tag>}, IsProxyRef>;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_swap_ranges/test.cpp:73]:
```suggestion
    test::CanCompare{derived_from<Category, forward_iterator_tag>}, IsProxyRef>;
```

File: tests/std/tests/P0896R4_ranges_alg_swap_ranges/test.cpp [hunk 3]
```diff
@@ -2,65 +2,105 @@
...
 #include <utility>
 
 #include <range_algorithm_support.hpp>
+using namespace std;
 
-constexpr void smoke_test() {
-    using ranges::swap_ranges, ranges::swap_ranges_result, ranges::iterator_t;
-    using std::same_as;
-    using I1 = iterator_t<basic_borrowed_range<int>>;
+// Validate that swap_ranges_result aliases in_out_result
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_swap_ranges/test.cpp:13]:
```suggestion
// Validate that swap_ranges_result aliases in_in_result
```

**Comment by StephanTLavavej** [comment]
Thanks for the ongoing modernizations! :rocket:
